shm_wrapper/shm_wrapper.c

```diff
@@ -422,57 +521,45 @@ No return value.
 */
 void device_read (int dev_ix, process_t process, stream_t stream, uint32_t params_to_read, param_val_t *params)
 {
+	char msg[64]; //for holding error message
+	
 	//check catalog to see if dev_ix is valid, if not then return immediately
 	if (!(shm_ptr->catalog & (1 << dev_ix))) {
-		printf("no device at this index, read failed\n");
+		sprintf(msg, "no device at dev_ix = %d, read failed", dev_ix);
+		log_runtime(DEBUG, msg);
 		return;
 	}
 	
-	//grab semaphore for the appropriate stream and device
-	if (stream == DATA) {
-		if (sem_wait(sems[dev_ix].data_sem) == -1) {
-			error("sem_wait: data sem @device_read");
-		}
-	} else {
-		if (sem_wait(sems[dev_ix].command_sem) == -1) {
-			error("sem_wait: command sem @device_read");
-		}
-	}
+	//call the helper to do the actual reading
+	device_read_helper(dev_ix, process, stream, params_to_read, params);
+}
+
+/*
+This function is the exact same as the above function, but instead uses the 64-bit device UID to identify
+the device that should be read, rather than the device index.
+*/
+void device_read_uid (uint64_t dev_uid, process_t process, stream_t stream, uint32_t params_to_read, param_val_t *params)
+{
+	char msg[64]; //for holding error message
+	int dev_ix = -1;
 	
-	//read all requested params
-	for (int i = 0; i < MAX_PARAMS; i++) {
-		if (params_to_read & (1 << i)) {
-			params[i] = shm_ptr->params[stream][dev_ix][i];
+	//check catalog and shm_ptr->dev_ids to determine if device exists
+	for (int i = 0; i < MAX_DEVICES; i++) {
+		if ((shm_ptr->catalog & (1 << i)) && (shm_ptr->dev_ids[i].uid == dev_uid)) {
```

Comment:
Would it be beneficial to have a map of uid -> idx because the executor will only be using these functions and I imagine the same for device handler?

Reply:
We could, but like it would require building / maintaining / searching an array each time, whereas this is just looping through an array so I feel like this is the faster solution? Unless I'm misunderstanding what you're saying.....

Reply:
I meant like a hashmap from device uid -> device idx. Since the UIDS are random, should be pretty spread hash buckets. But that might be a lot of work in C and the array size is fixed at 32 anyways. I think this is fine.

Reply:
ok sounds good, I mean this could be something we investigate later, it might be possible to get some speedup with this. But for now I think it's fine, like you said~~